File: Chapter 03/Python/section_3_2_coupling_exercise/before/user_report_generator.py
from .user_data_service import UserDataService

"""
ANTI-PATTERN: The Tightly Coupled Client.
* ARCHITECTURE PROBLEM: Because the service is too granular, this generator 
is forced to act as an orchestrator. 
* The generator now knows too much about the internal data structure (User -> 
Orders -> Totals). If the service changes how it stores orders, this 
generator will break. This is a classic violation of Encapsulation.
"""

class UserReportGenerator:
    def __init__(self):
        self.data_service = UserDataService()

    def generate_report(self, user_id):
        # 🚨 ARCHITECTURE WARNING: High temporal coupling. The client must 
        # manage the order of execution for these specific methods.
        name = self.data_service.get_user_name(user_id)
        email = self.data_service.get_user_email(user_id)
        orders = self.data_service.get_user_order_ids(user_id)

        total_spent = 0.0
        
        # 🚨 ARCHITECTURE WARNING: "Chatty" execution.
        # The client is performing business logic (summing totals) that 
        # should belong to the data service.
        for order_id in orders:
            total_spent += self.data_service.get_order_total(order_id)

        return f"User Report for {name} ({email}) - Total Spent: ${total_spent:.2f}"

File: Chapter 03/JS/section_3_2_coupling_exercise/before/userReportGenerator.js
const UserDataService = require('./userDataService');

/**
 * ANTI-PATTERN: The Tightly Coupled Client.
 * * ARCHITECTURE PROBLEM: Because the UserDataService is too granular, this client 
 * class is forced to take on the responsibility of orchestrating the data gathering. 
 * * The generator now knows WAY too much about the internal structure of the data. 
 * It knows that it has to fetch the user, then fetch the orders, and then loop 
 * through the orders to calculate a total. This is a severe violation of encapsulation.
 */
class UserReportGenerator {
    constructor() {
        this.dataService = new UserDataService();
    }

    generateReport(userId) {
        
        // 🚨 ARCHITECTURE WARNING: High temporal coupling. The client has to 
        // call these specific methods in a specific sequence to get what it needs.
        const name = this.dataService.getUserName(userId);
        const email = this.dataService.getUserEmail(userId);
        const orders = this.dataService.getUserOrderIds(userId);

        let totalSpent = 0;
        
        // 🚨 ARCHITECTURE WARNING: Feature Envy / Chatty Execution.
        // The client is forced to loop through orders and request totals one by one.
        // It is doing the heavy lifting that the service should be doing for it!
        for (const orderId of orders) {
            totalSpent += this.dataService.getOrderTotal(orderId);
        }

        return `User Report for ${name} (${email}) - Total Spent: $${totalSpent.toFixed(2)}`;
    }
}

module.exports = UserReportGenerator;